core/shared/src/main/scala/coursier/util/Parse.scala

```diff
@@ -191,59 +206,81 @@ object Parse {
           })
 
       case Array(org, "", rawName, version) =>
         module(s"$org::$rawName", defaultScalaVersion)
           .right
           .map(mod => {
-            Dependency(
-              mod,
-              version,
-              configuration = defaultConfig,
-              attributes = attributes,
-              transitive = transitive,
-              exclusions = localExcludes.getOrElse(mod.orgName, Set()) | globalExcludes)
+            (Dependency(
+                mod,
+                version,
+                configuration = defaultConfig,
+                attributes = attributes,
+                transitive = transitive,
+                exclusions = localExcludes.getOrElse(mod.orgName, Set()) | globalExcludes),
+              extraDependencyParams)
           })
 
       case Array(org, rawName, version, config) =>
         module(s"$org:$rawName", defaultScalaVersion)
           .right
           .map(mod => {
-            Dependency(
-              mod,
-              version,
-              config,
-              attributes,
-              transitive = transitive,
-              exclusions = localExcludes.getOrElse(mod.orgName, Set()) | globalExcludes)
+            (Dependency(
+                mod,
+                version,
+                config,
+                attributes,
+                transitive = transitive,
+                exclusions = localExcludes.getOrElse(mod.orgName, Set()) | globalExcludes),
+              extraDependencyParams)
           })
 
       case Array(org, rawName, version) =>
         module(s"$org:$rawName", defaultScalaVersion)
           .right
           .map(mod => {
-            Dependency(
-              mod,
-              version,
-              configuration = defaultConfig,
-              attributes = attributes,
-              transitive = transitive,
-              exclusions = localExcludes.getOrElse(mod.orgName, Set()) | globalExcludes)
+            (Dependency(
+                mod,
+                version,
+                configuration = defaultConfig,
+                attributes = attributes,
+                transitive = transitive,
+                exclusions = localExcludes.getOrElse(mod.orgName, Set()) | globalExcludes),
+              extraDependencyParams)
           })
 
       case _ =>
         Left(s"Malformed dependency: $s")
     }
   }
 
+  /**
+   * Validates the parsed attributes
+   *
+   * Currently only "classifier" and "url" are allowed. If more are
+   * added, they should be passed in via the second parameter
+   *
+   * @param attrs Attributes parsed
+   * @param validAttrsKeys Valid attribute keys
+   */
+  private def validateAttributes(attrs: Map[String, String], validAttrsKeys: Set[String]): Unit = {
+    val extraAttributes = attrs.keys.toSet.diff(validAttrsKeys)
+
+    if (attrs.size > validAttrsKeys.size || extraAttributes.nonEmpty)
+      throw new ModuleParseError(s"The only attributes allowed are: ${validAttrsKeys.mkString(", ")}. ${
```

Comment:
I'll probably deprecate `ModuleParseError` in a later PR. In the public methods of this file, errors should be returned via a `Left("error message")` rather than an exception. (I guess I underlooked it when it was added :| )

Reply:
my bad :)